### ISSUE TITLE
Time delay on nodes

### ISSUE THREAD

**Issue by snorfalorpagus** [pywr/pywr#505]
We discussed adding functionality for time delays in flows in the network. This would need to be done somewhere as a cross between the Storage and Catchment nodes - a break in the route, and an Input fixed to the flow of an Output but with X days delay in timestep.

**Comment by Batch21** [comment]
I've stated looking at this and have something that seems to work:

```python

class DelayNode(Node):

    def __init__(self, model, name, delay, **kwargs):
        super().__init__(model, name, **kwargs)
        self.allow_isolated = True
        output_name = "{} - output".format(name)
        input_name = "{} - input".format(name)
        assert(output_name not in model.nodes)
        assert(input_name not in model.nodes)

        self.output = Output(model, name=output_name)
        self.delay_param = FlowDelayParameter(model, self.output, delay, name="delay_param")
        self.input =  Catchment(model, name=input_name, flow=self.delay_param)

    def iter_slots(self, slot_name=None, is_connector=True):
        if is_connector:
            yield self.input
        else:
            yield self.output


class FlowDelayParameter(Parameter):

    def __init__(self, model, node, delay, *args, **kwargs):
        super().__init__(model, *args, **kwargs)
        self.node = node
        self.delay = delay

    def setup(self):
        super().setup()
        ncombs = len(self.model.scenarios.combinations)
        self._memory = np.zeros((self.delay, ncombs))

    def value(self, ts, scenario_index):
        return self._memory[-1, scenario_index.global_id]

    def after(self):
        self._memory = np.roll(self._memory, 1)
        self._memory[0, :] = self.node.flow[:]

```

@jetuk you mentioned a delay parameter in #2 . Is the above anything like you imagined?

**Comment by jetuk** [comment]
Yes, I think this is in principle how this would work. Couple of comments:

- I would use `Input` instead of `Catchment` and explicitly set `min_flow` and `max_flow` to be the same thing.
- You may want to set the flow of `DelayNode` to equal either the incoming or outgoing flow so that it returns some values if a recorder is looking at it.
- We could support setting delay in terms of number of timesteps or number of days. See approach on `RollingMeanFlowNodeRecorder`.

**Comment by Batch21** [comment]
> Yes, I think this is in principle how this would work. Couple of comments:
> 
> * I would use `Input` instead of `Catchment` and explicitly set `min_flow` and `max_flow` to be the same thing.
> * You may want to set the flow of `DelayNode` to equal either the incoming or outgoing flow so that it returns some values if a recorder is looking at it.
> * We could support setting delay in terms of number of timesteps or number of days. See approach on `RollingMeanFlowNodeRecorder`.

Thanks. I'll update based on your comments, finish off some tests and submit as a PR

**Comment by jetuk** [comment]
Ideally Cythonise the `FlowDelayParameter` parameter for performance reasons.

**Comment by jetuk** [comment]
Fixed in #904
